native/thrift_message_read_module.c

```diff
@@ -0,0 +1,338 @@
+
+    }
+    if (!strcmp(proto, "TCompactProtocol")) {
+        int ver_type;
+        int version;
+        int state;
+        PyObject *py_state;
+
+        py_state = PyDict_GetItemString(globalVars,"state");
+        if (!PyArg_Parse(py_state, "i", &state)) {
+            return ;
+        }
+        if (state != CLEAR) {
+            PyErr_SetString(PyExc_ValueError, "Bad state");
+            return NULL;
+        }
+        int proto_id = readByteC(data);
+        if (proto_id != COMPACT_PROTOCOL_ID) {
+            PyErr_SetString(PyExc_ValueError, "Bad protocol ID");
+            return NULL;
+        }
+        ver_type = readByteC(data);
+        *type = (ver_type>>TYPE_SHIFT_AMOUNT) & TYPE_BITS;
+        version = ver_type & COMPACT_VERSION_MASK;
+        if (version != COMPACT_VERSION) {
+             PyErr_SetString(PyExc_ValueError,
+                            "Bad compact message version");
+             return NULL;
+        }
+        *seqid = readVarint(data);
+        return readBinary(data, proto);
+    }
+    return NULL;
+}
+
+static void
+readMessageEnd(char *data, char *proto)
+{
+    if (!strcmp(proto, "TJSONProtocol")) {
+        readJSONArrayEnd(data);
+    }
+    else if (!strcmp(proto, "TCompactProtocol")) {
+        int state;
+        PyObject *globalVars;
+        PyObject *py_state;
+        PyObject *py_size;
+        Py_ssize_t containersTop;
+
+        globalVars = PyThreadState_GetDict();
+        py_size = PyDict_GetItemString(globalVars, "containersStackTop");
+        if (!PyArg_Parse(py_size, "n", &containersTop)) {
+                return ;
+        }
+        py_state = PyDict_GetItemString(globalVars,"state");
+        if (!PyArg_Parse(py_state, "i", &state)) {
+            return ;
+        }
+        if (state != CLEAR) {
+            PyErr_SetString(PyExc_ValueError, "Wrong State");
+            return ;
+        }
+        if (containersTop != -1) {
+            PyErr_SetString(PyExc_ValueError,
+                            "structs stack is not empty");
+            return ;
+        }
+    }
+}
+
+static char *
+message_type_to_str(int mtype)
+{
+        char *messageType;
+        if (mtype == CALL) {
+            messageType = malloc(sizeof(char)*(strlen("call")+1));
+            strcpy(messageType, "call");
+            return messageType;
+        }
+        else if (mtype == REPLY) {
+            messageType = malloc(sizeof(char)*(strlen("reply")+1));
+            strcpy(messageType, "reply");
+            return messageType;
+        }
+        else if (mtype == EXCEPTION) {
+            messageType = malloc(sizeof(char)*(strlen("exception")+1));
+            strcpy(messageType, "exception");
+            return messageType;
+        }
+        else if (mtype == ONEWAY) {
+            messageType = malloc(sizeof(char)*(strlen("oneway")+1));
+            strcpy(messageType, "oneway");
+            return messageType;
+        }
+        else {
+            return NULL;
+        }
+}
```

Comment:
when returning constant strings it's better to avoid malloc(), i.e.:

https://gist.github.com/rgs1/8928c8751042274c4707a24f5b5a918e